system/security/cryptohelper/pke.cpp

```diff
@@ -0,0 +1,245 @@
+
+    /* Provide the message to be encrypted, and obtain the encrypted output.
+     * EVP_EncryptUpdate can be called multiple times if necessary
+     */
+
+    const size32_t cipherBlockSz = 128;
+    size32_t outMaxSz = inSz + cipherBlockSz/8;
+    byte *outPtr = (byte *)out.reserveTruncate(outMaxSz);
+    int outSz;
+    if (1 != EVP_EncryptUpdate(ctx, (unsigned char *)outPtr, &outSz, (unsigned char *)inBytes, inSz))
+        throw makeEVPException(0, "Failed EVP_EncryptUpdate");
+    int ciphertext_len = outSz;
+
+    /* Finalise the encryption. Further ciphertext bytes may be written at
+     * this stage.
+     */
+    if (1 != EVP_EncryptFinal_ex(ctx, outPtr + outSz, &outSz))
+        throw makeEVPException(0, "Failed EVP_EncryptFinal_ex");
+    ciphertext_len += outSz;
+    out.setLength(ciphertext_len);
+}
+
+void aesKeyDecrypt(MemoryBuffer &out, size32_t inSz, const void *inBytes, const char *key, const char *iv)
+{
+    OwnedCipherCtx ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        throw makeEVPException(0, "Failed EVP_CIPHER_CTX_new");
+
+    const size32_t cipherBlockSz = 128;
+    // man page says(!) - "should have sufficient room for (inl + cipher_block_size) bytes unless the cipher block size is 1 in which case inl bytes is sufficient
+    size32_t outMaxSz = ((inSz==1?8:inSz) + cipherBlockSz/8);
+    byte *outPtr = (byte *)out.reserveTruncate(outMaxSz);
+
+    /* Initialise the decryption operation. IMPORTANT - ensure you use a key
+     * and IV size appropriate for your cipher
+     * In this example we are using 256 bit AES (i.e. a 256 bit key). The
+     * IV size for *most* modes is the same as the block size. For AES this
+     * is 128 bits
+     * */
+    if (1 != EVP_DecryptInit_ex(ctx, EVP_aes_256_cbc(), nullptr, (const unsigned char *)key, (const unsigned char *)iv))
+        throw makeEVPException(0, "Failed EVP_EncryptInit_ex");
+
+    /* Provide the message to be decrypted, and obtain the plaintext output.
+     * EVP_DecryptUpdate can be called multiple times if necessary
+     */
+    int outSz;
+    if (1 != EVP_DecryptUpdate(ctx, outPtr, &outSz, (const unsigned char *)inBytes, inSz))
+        throw makeEVPException(0, "Failed EVP_DecryptUpdate");
+    int plaintext_len = outSz;
+
+    /* Finalise the decryption. Further plaintext bytes may be written at
+     * this stage.
+     */
+    if (1 != EVP_DecryptFinal_ex(ctx, outPtr + outSz, &outSz))
+        throw makeEVPException(0, "Failed EVP_DecryptFinal_ex");
+
+    plaintext_len += outSz;
+    out.setLength(plaintext_len);
+}
+
+// RSA ...
+void publicKeyEncrypt(MemoryBuffer &out, size32_t inSz, const void *inBytes, const CLoadedKey &publicKey)
+{
+    OwnedEVPPkeyCtx ctx = EVP_PKEY_CTX_new(publicKey, nullptr);
+    if (!ctx || (EVP_PKEY_encrypt_init(ctx) <= 0) || (EVP_PKEY_CTX_set_rsa_padding(ctx, RSA_PKCS1_PADDING)) <= 0)
+        throwEVPException(0, "publicKeyEncrypt: failed to initialize key: %s", publicKey.queryKeyName());
```

Comment:
It will be hard to determine which of either EVP_PKEY_encrypt_init or EVP_PKEY_CTX_set_rsa_padding failed without dropping into assembler to see which one returns <=0. Perhaps break them up to separate lines?

Reply:
not really, because throwEVPException will contain the text, containing the code to the function and the reason.